src/chain/shared_state.rs

```diff
@@ -156,6 +157,35 @@ impl SharedState {
                 .collect()
         }
     }
+
+    /// Updates the entry in `their_keys` for `prefix` to the latest known key; if a split
+    /// occurred in the meantime, the keys for sections covering the rest of the address space are
+    /// initialised to the old key that was stored for their common ancestor
+    pub fn update_their_keys(&mut self, prefix: Prefix<XorName>, key: BlsPublicKey) {
+        if let Some(&pfx) = self
+            .their_keys
+            .keys()
+            .find(|pfx| pfx.is_compatible(&prefix))
+        {
+            let old_key = unwrap!(self.their_keys.remove(&pfx));
+            let mut current_pfx = prefix;
+            loop {
+                current_pfx = current_pfx.sibling();
+                if self.their_keys.contains_key(&current_pfx) || self.our_prefix() == &current_pfx {
+                    break;
+                }
+                let _ = self.their_keys.insert(current_pfx, old_key.clone());
+                current_pfx = current_pfx.popped();
```

Comment:
I'm not 100% happy with the `current_pfx` modified twice. Also it seem more of  while loop.
```
let mut current_pfx = prefix.sibling();
while !self.their_keys.contains_key(&current_pfx) && self.our_prefix() != &current_pfx {
     let _ = self.their_keys.insert(current_pfx, old_key.clone());
     current_pfx = current_pfx.popped().sibling();
}
```

It can probably be made still clearer.